Move node launching to network layer

rvl opened the pull request:
This adds a `withNetworkLayer` function which handles launching the node.

- Fixes package dependency issue (stack exec in tests)

- Fixes code coverage difficulties associated with relaunching executable.

- Removes need to export JormungandrLayer implementation details from network layer, and so reduces the number of constructors exported.

- Allows all network layer setup to be done in the network layer and not duplicated at each call site.

- Fixes "The assertion in this test case is wrong" pending test cases.

- Reduces amount that needs testing -- launcher scenarios being duplicates of server scenarios. Although no tests have been removed yet.

- Removes proxying of command line arguments from launcher to server.

- Same process supervision capability as before, except with one less process. Same user convenience as before, with less maintenance burden for us.

- Adds handling for the error case where the node executable is not found. Adds structured logging to the launcher module.

- We can run a NetworkLayer without a WalletLayer, but still have launch functionality.

- Fix non-portable and unbracketed use of `/tmp/jormungandr` in tests.

- Improve db file cleanup in acceptance test.

- In theory, the tests can use more of the application code to start the wallet (and/or components) rather than their own test-specific code.

### Comments

To make review easier, just skim the http-bridge changes.

rvl committed:
Jormungandr.NetworkSpec: Remove use of generic-arbitrary

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -740,11 +720,17 @@ randomPortOption = flag' False $ mempty
     <> metavar "INT"
     <> help "port used for communicating with the target node."
-    <> value (Port 8080)
     <> showDefaultWith showT
```
minor: `showDefaultWith` now pointless

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -439,7 +438,7 @@ dbLayer =
     :: forall t ctx. HasGenesisData t ctx
     => Lens' ctx (Block (Tx t), BlockchainParameters)
-genesisData =
+genesisData = -- networkLayer . staticBlockchainParameters
```
:thinking:  ?

rvl replied:
These values can be gotten from from the network layer. But I removed the comment. Maybe it's better that the wallet has these values independently of the network layer.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -242,11 +243,26 @@ withDBLayer
+    -> ((SqliteContext, DBLayer IO s t k) -> IO a)
+       -- ^ Action to run.
+    -> IO a
+withDBLayerCtx logConfig trace fp = bracket before after
```
I'd rather have only one `withDBLayer` that does more :+1:
Yet, there's a legitimate question of whether the SqlContext should be accessible outside of the DBLayer :thinking: ? 

Not saying we should get rid of this _now_, but that's something to consider cleaning up later.

Anviking commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -9,27 +8,26 @@ module Cardano.Wallet.Network
 import Prelude
 
+-- import Cardano.Wallet.Primitive.Chain
+--     ( BlockchainParameters (..) )
```
Commented import

rvl replied:
I was thinking `Cardano.Wallet.Primitive.Chain` might be a possible new place to put BlockchainParameters.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -439,7 +438,7 @@ dbLayer =
     :: forall t ctx. HasGenesisData t ctx
     => Lens' ctx (Block (Tx t), BlockchainParameters)
-genesisData =
+genesisData = -- networkLayer . staticBlockchainParameters
```
A comment

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -596,7 +614,8 @@ checkpointFromEntity cp utxo s =
     blockHeight' = Quantity . toEnum . fromEnum $ bh
     bp = W.BlockchainParameters
-        { getGenesisBlockDate = W.StartTime genesisStart
+        { getGenesisBlockHash = coerce genesisHash
```
Wait, do we really need this? We could still extract it from the `NetworkLayer`, but not have it in `BlockchainParameters` and the DB?

I guess it could theoretically be nice keep it together with the other data in the checkpoints though to make sure they are consistent.

rvl replied:
At the moment all of the stuff in BlockchainParameters is effectively static because it comes from the genesis block, and there is no way to get parameter updates. Genesis block date, hash, and slot duration will never change.

The genesis hash is needed to instantiate the TransactionLayer.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -188,7 +191,8 @@ instance NFData BlockchainParameters
 instance Buildable BlockchainParameters where
     build bp = blockListF' "" id
-        [ "Genesis block date: " <> startTimeF (getGenesisBlockDate bp)
+        [ "Genesis block hash: " -- <> build (getGenesisBlockHash bp)
```
Commented code

rvl replied:
Fixed now

rvl committed:
Move node launching to network layer

- Fixes package dependency issue (stack exec in tests)

- Fixes code coverage difficulties associated with relaunching
  executable.

- Removes need to export JormungandrLayer implementation details from
  network layer, and so reduces the number of constructors exported.

- Allows all network layer setup to be done in the network layer and
  not duplicated at each call site.

- Fixes "The assertion in this test case is wrong" pending test cases.

- Reduces amount that needs testing -- launcher scenarios being
  duplicates of server scenarios. Although no tests have been removed
  yet.

- Removes proxying of command line arguments from launcher to server.

- Same process supervision capability as before, except with one less
  process. Same user convenience as before, with less maintenance burden
  for us.

- Adds handling for the error case where the node executable is not
  found. Adds structured logging to the launcher module.

- We can run a NetworkLayer without a WalletLayer, but still have
  launch functionality.

- Fix non-portable and unbracketed use of /tmp/jormungandr in tests.

- Improve db file cleanup in acceptance test.

- In theory, the tests can use more of the application code to start
  the wallet (and/or components) rather than their own test-specific
  code.

- Test faucet doesn't need to get block0 hash because that's handled
  by network layer init.

KtorZ commented on lib/core/src/Cardano/Wallet/Network/Ports.hs:
```diff
@@ -0,0 +1,135 @@
+-- test suites running in parallel.
+--
+
+module Cardano.Wallet.Network.Ports
```
does this replace the `test-utils` package ?

rvl replied:
Unfortunately I can't get rid of the Ports module in test-utils because of a circular dependency.

Anviking commented on lib/jormungandr/exe/cardano-wallet-jormungandr.hs:
```diff
@@ -294,91 +245,113 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
-                , "but there's no initial fee policy defined?"
-                ]
+serveWallet
+    :: forall t k n s. (n ~ 'Testnet, t ~ Jormungandr n, s ~ SeqState t, k ~ SeqKey)
```
This line is wider than 80 characters. I think a few other in this PR are too.

rvl replied:
Fixed thanks.

Anviking commented on lib/jormungandr/exe/cardano-wallet-jormungandr.hs:
```diff
@@ -294,91 +245,113 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
             exitFailure
+        ErrStartupCommandExited pe -> case pe of
+            ProcessDidNotStart _cmd exc -> do
+                logAlert tr $ "Could not start the node backend. " <> T.pack (show exc)
```
Looks nice!

Anviking commented on lib/jormungandr/test/integration/Main.hs:
```diff
@@ -132,32 +129,35 @@ main = hspec $ do
-                Sqlite.destroyDBLayer db
-                threadDelay oneSecond
-        return $ Context cleanup manager wPort jPort faucet estimator Proxy
+    withServer :: Maybe Bool -> IO () -> (Context (Jormungandr 'Testnet) -> IO a) -> IO a
```
How does this `Maybe Bool` work?

It seems to me like `Nothing` results in a random port.

https://github.com/input-output-hk/cardano-wallet/blob/d0d56f9c91448b4ae4ab74ec0ac65f064286c410/lib/jormungandr/test/integration/Main.hs#L188

And it seems like the other two cases also result in a random port?

rvl replied:
Yeah I don't know. I just replicated how it was before.

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs:
```diff
@@ -89,9 +112,56 @@ import qualified Data.ByteString.Lazy as BL
+    :: forall n a. KnownNetwork n
+    => Trace IO Text
+    -> HttpBridgeBackend
+    -> (PortNumber -> Either ErrStartup (NetworkLayer IO Tx (Block Tx)) -> IO a)
```
Shouldn't `PortNumber` be part of the `Right` branch on the either here? (the fact that we give a default value of `0` in case of error is a good hint to me..)

rvl replied:
Yes it should, fixed.

Anviking commented on lib/jormungandr/test/integration/Main.hs:
```diff
@@ -108,7 +105,7 @@ instance KnownCommand (Jormungandr n) where
 
 main :: forall t. (t ~ Jormungandr 'Testnet) => IO ()
-main = hspec $ do
+main = hspec $ parallel $ do
```
So we are running multiple `jörmungandr` at the same time, with random ports and random state-directories? Maybe worth clarifying what `withCardanoWalletServer` or `withServer` does.

rvl replied:
Yes, I fixed it. `parallel` is unlikely to work with the E2E tests.

Anviking commented on lib/launcher/src/Cardano/Launcher.hs:
```diff
@@ -89,24 +129,96 @@ instance Buildable Command where
+    sleep = forever $ threadDelay maxBound
+    start = async . flip (withBackendProcess tr) sleep
+
+withBackendProcess :: Trace IO LauncherLog -> Command -> IO a -> IO (Either ProcessHasExited a)
```
What is the purpose of this? A short doc-comment? 🙏 

It combines logging with launching of the backend-process, and is used by the specific targets? Anything more?

rvl replied:
I added a doc string. The important thing is that it is a bracket-style function for handling clean up of resources.

Anviking commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/Launch.hs:
```diff
@@ -8,60 +8,56 @@
+        (True, _) -> putStrLn $ "Not cleaning up temporary directory " ++ d
+        (_, True) -> removeDirectoryRecursive d
+        _ -> pure ()
+
```
```suggestion
-- | Launches jörmungandr, but not the wallet
```
??? I think it could be more clear what this is doing. Also what is the purpose? It seems unused.

rvl replied:
I thought it would be needed in the tests but perhaps not.

Anviking reviewed (approved):
Seems like many a good things. What I've been able to comprehend looks largely good to me.

I _was_ hoping to make the `NetworkLayer` smaller. Making it bigger seems to work pretty well though, as you show here.

There are a few places where I think doc-comments could clarify what is going on (see comments). Maybe some module level doc-comment could benefit from being updated too, not sure.

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge.hs:
```diff
@@ -0,0 +1,156 @@
+            Right nl -> do
+                wl <- newWalletLayer db nl
+                withServer wl $ \apiPort -> do
+                    let action = fromMaybe (\_ _ _ _ _ -> forever $ threadDelay maxBound) mAction
```
I am sure we can do better here :grimacing: 
What about passing the `maybe action` down to `withServer` and have a pattern match here:

- If there's an action, we go for the double-race
- If not, then we only spawn the wallet server until it dies.

rvl replied:
Yeah it seems like whenever i start using `Maybe (IO a)` the code gets ugly.
I have fixed it to pass the `maybe action` down.

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge.hs:
```diff
@@ -0,0 +1,156 @@
+    -- ^ HTTP API Server port.
+    -> HttpBridgeBackend
+    -- ^ Whether and how to launch or use the node backend.
+    -> Maybe (PortNumber -> NetworkLayer IO Tx (Block Tx) -> Sqlite.SqliteContext -> DBLayer IO s t k -> WalletLayer s t k -> IO a)
```
This `SqliteContext` argument looks off / awkward to me here.

rvl replied:
Yeah unfortunately I can't get rid of it yet because this callback is used for tests and some of the tests want SqliteContext.

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs:
```diff
@@ -100,15 +170,16 @@ mkNetworkLayer httpBridge = NetworkLayer
 newNetworkLayer
     :: forall n. KnownNetwork (n :: Network)
-    => Int
+    => PortNumber
```
Minor suggestion, but we have a `Port (purpose :: Symbol)` type somewhere.. I think it's in the `cardano-wallet-cli` module. It's quite handy for self-documenting signature, especially because it gets quite confusing quite quickly which port is what (backend API vs our API)

KtorZ reviewed (approved):
It's good that we are finally biting the bullet here! 
I think we can make some small improvements in type signatures to avoid awkward default cases in a few places. 
Also, this exposed `SqliteContext` bother me somehow but that's probably something we can try to get rid off later.

I am approving, yet quite a few areas also need some revision regarding our style guide (I didn't pointed that out in the review but one-liner signatures mostly that could be written on several lines for better readability).

Nice piece of work. Let's not keep that PR open for too long as it will degrades quickly and make synchronization with other PRs quite hard.

rvl committed:
Jormungandr: Fix launcher test

Fixed a few logging tests.

Some of the tests still want port 8090, annoying.

KtorZ committed:
fix some integration tests failures in Jörmungandr

KtorZ committed:
fix log line order in http-bridge executable

KtorZ committed:
move 'JormungandrConnParams' inside the success part of the callback

KtorZ committed:
Jormungandr: use 'serveWallet' in integration tests

avoiding logic duplication and greatly simplifying the integration setup

KtorZ committed:
update http-bridge after moving 'waitForConnection' back to Cardano.Wallet.Network

KtorZ committed:
regenerate nix machinery
